### PR TITLE
Bring `_collections_abc` closer to runtime definition

### PR DESCRIPTION
**Pull request by AlexWaygood**
Currently, the following code succeeds at runtime, but [fails mypy](https://mypy-play.net/?mypy=latest&python=3.10&flags=show-error-codes&gist=4671377be6f6e4fa6601f13e41cdfbb4):

```python
from _collections_abc import dict_keys
```

This PR proposes several changes to bring the `_collections_abc` stub closer to its runtime definition. The significant change is moving the definitions of `dict_keys`, `dict_values` and `dict_items` from `builtins` to `_collections_abc`.

I have also added `@final` to these classes, as they cannot be subclassed at runtime, and added `# type: ignore` comments to `_OrderedDictKeysView`/`_OrderedDictValuesView`/`_OrderedDictItemsView`. (These are indeed subclasses of `dict_keys`/`dict_values`/`dict_items`, even though the three classes cannot be subclassed _at runtime_; see #6299.)

### CODE REVIEW

**Comment by AlexWaygood** [comment]
@srittau, it looks like #6252 might be causing CI failures on unrelated PRs? 🙂

**Comment by srittau** [comment]
The problem wasn't #6252, but a new release of fpdf2, which added arguments. Anyway, #6313 will fix CI.

File: stdlib/_collections_abc.pyi [hunk 1]
```diff
@@ -1,5 +1,8 @@
+import sys
+from types import FunctionType as FunctionType, MappingProxyType as mappingproxy  # undocumented
```
Comment (JelleZijlstra) [inline stdlib/_collections_abc.pyi:2]: If they're undocumented, why are we re-exporting them here?

Comment (AlexWaygood) [reply stdlib/_collections_abc.pyi:2]: I was going according to the principle of keeping the stubs as close as possible to the runtime definition; `from _collections_abc import mappingproxy` works fine at runtime, even if it's a *pretty weird* thing to be doing. But, definitely don't want to die on this hill! I'll take them out if you think it's unnecessary.

Comment (AlexWaygood) [reply stdlib/_collections_abc.pyi:2]: Cool, I'll remove them

Comment (AlexWaygood) [reply stdlib/_collections_abc.pyi:2]: Best to remove `generator`, `coroutine` and `async_generator` as well, do you think?

Comment (JelleZijlstra) [reply stdlib/_collections_abc.pyi:2]: Yes

**Comment by KotlinIsland** [comment]
`_collections_abc.dict_keys` is generic in the stubs here, but it's not generic in reality, such that:

```py
from _collections_abc import dict_keys

dict_keys[int, int]
```

Will type check fine, but raise errors at runtime.

**Comment by AlexWaygood** [comment]
> `_collections_abc.dict_keys` is generic in the stubs here, but it's not `Generic` in reality

I agree that this is unfortunate, but I'm not entirely sure  what your point is. `dict_keys` was generic before this PR, so this isn't a new change.

Yes it's not great that the stub differs from the runtime implementation in this case, and yes the stub should generally be as close to the runtime implementation as possible. But having these classes be generic in the stub file allows type checkers to perform far more sophisticated type inferences.

If you have a valid use case for wanting to parameterize `dict_keys` at runtime, like it is in the stub, you can always make a feature request at bugs.python.org! Personally, I can't think of a case where I'd want or need to, however 🙂

**Comment by KotlinIsland** [comment]
Previously this definition was `_dict_keys` in builtins so it was a lot harder to get at, but I completely agree with you. Maybe something like a `@StubOnlyGeneric` annotation could be added.

**Pull request merged**